src/webui/api/searchcontroller.cpp

```diff
@@ -0,0 +1,384 @@
+        categories << SearchPluginManager::categoryFullName(category);
+
+    const QJsonArray result = QJsonArray::fromStringList(categories);
+    setResult(result);
+}
+
+void SearchController::pluginsAction()
+{
+    const QStringList allPlugins = SearchPluginManager::instance()->allPlugins();
+    setResult(getPluginsInfo(allPlugins));
+}
+
+void SearchController::installPluginAction()
+{
+    checkParams({"sources"});
+
+    const QStringList sources = params()["sources"].split('|');
+    for (const QString &source : sources)
+        SearchPluginManager::instance()->installPlugin(source);
+}
+
+void SearchController::uninstallPluginAction()
+{
+    checkParams({"names"});
+
+    const QStringList names = params()["names"].split('|');
+    for (const QString &name : names)
+        SearchPluginManager::instance()->uninstallPlugin(name.trimmed());
+}
+
+void SearchController::enablePluginAction()
+{
+    checkParams({"names", "enable"});
+
+    const QStringList names = params()["names"].split('|');
+    const bool enable = Utils::String::parseBool(params()["enable"].trimmed(), false);
+
+    for (const QString &name : names)
+        SearchPluginManager::instance()->enablePlugin(name.trimmed(), enable);
+}
+
+void SearchController::updatePluginsAction()
+{
+    SearchPluginManager *pluginManager = SearchPluginManager::instance();
+
+    connect(pluginManager, &SearchPluginManager::checkForUpdatesFinished, this, &SearchController::checkForUpdatesFinished);
+    connect(pluginManager, &SearchPluginManager::checkForUpdatesFailed, this, &SearchController::checkForUpdatesFailed);
+    pluginManager->checkForUpdates();
+}
+
+void SearchController::searchTimeout(const QString &id)
+{
+    ISession *session = sessionManager()->session();
+    auto searchHandlers = session->getData(QLatin1String("searchHandlers")).value<SearchHandlerDict>();
+
+    if (searchHandlers.contains(id)) {
+        LogMsg(tr("Deleting search session %1 from web session due to timeout").arg(id), Log::INFO);
+        SearchHandlerPtr searchHandler = searchHandlers[id];
+        searchHandler->cancelSearch();
+        searchHandlers.remove(id);
+        session->setData(QLatin1String("searchHandlers"), QVariant::fromValue(searchHandlers));
+
+        auto activeSearches = session->getData(QLatin1String("activeSearches")).value<QSet<QString>>();
+        activeSearches.remove(id);
+        session->setData(QLatin1String("activeSearches"), QVariant::fromValue(activeSearches));
+    }
+}
+
+void SearchController::checkForUpdatesFinished(const QHash<QString, PluginVersion> &updateInfo)
+{
+    if (updateInfo.isEmpty()) {
+        LogMsg(tr("All plugins are already up to date."), Log::INFO);
+        return;
+    }
+
+    LogMsg(tr("Updating %1 plugins").arg(updateInfo.size()), Log::INFO);
+
+    SearchPluginManager *pluginManager = SearchPluginManager::instance();
+    for (const QString &pluginName : updateInfo.keys()) {
```

Comment:
Consider 
```c++
for (const auto &plugin : updateInfo) {
    pluginManager->updatePlugin(plugin.key());
}
```

Reply:
What type of `plugin` do you expect?

Reply:
Oh, sorry. Stupid Qt dictionaries...

Reply:
@glassez, what exactly is wrong here? You have to copy keys into a dedicated container to iterate over them. Is it not stupid?

Reply:
Why don't use another way to iterate over it?

Reply:
Return to using iterators explicitly? Or what did you suggest?

Reply:
>Return to using iterators explicitly?

Why not? Or is it no longer fashionable?

Reply:
Sure, everyone fancies the range based for loop. Explicit iterators are so old-styled.

Reply:
IMO, there is no place for fashion in programming. We should select the appropriate tool in each case.